File: apps/android/app/src/main/java/com/karibuhealth/app/domain/InpatientIvCatalog.kt
package com.karibuhealth.app.domain

import java.time.Instant

/** IV fluids and additives common at Ugandan HC III. */
object InpatientIvCatalog {

    data class FluidOption(val id: String, val label: String)

    val fluids = listOf(
        FluidOption("normal_saline", "Normal saline 0.9%"),
        FluidOption("ringers_lactate", "Ringer's lactate"),
        FluidOption("d5", "Dextrose 5%"),
        FluidOption("d10", "Dextrose 10%"),
        FluidOption("half_normal", "Half-normal saline 0.45%"),
        FluidOption("d5_ns", "D5 in normal saline"),
    )

    val additives = listOf(
        FluidOption("none", "None (fluid only)"),
        FluidOption("vit_b_complex", "Vitamin B complex"),
        FluidOption("vit_c", "Vitamin C"),
        FluidOption("ceftriaxone", "Ceftriaxone"),
        FluidOption("metronidazole", "Metronidazole"),
        FluidOption("artesunate", "Artesunate"),
        FluidOption("quinine", "Quinine"),
        FluidOption("oxytocin", "Oxytocin (PPH)"),
        FluidOption("magnesium_sulphate", "Magnesium sulphate"),
    )

    val volumePresetsMl = listOf(500, 1000, 2000)

    fun fluidLabel(id: String): String = fluids.find { it.id == id }?.label ?: id

    fun additiveLabel(id: String?): String? {
        if (id.isNullOrBlank() || id == "none") return null
        return additives.find { it.id == id }?.label ?: id
    }

    fun estimateMlRemaining(volumeMl: Int, rateMlHr: Int?, startedAt: String, now: Long = System.currentTimeMillis()): Int? {
        if (rateMlHr == null || rateMlHr <= 0) return null
        val elapsedHr = (now - Instant.parse(startedAt).toEpochMilli()) / 3_600_000.0
        return (volumeMl - elapsedHr * rateMlHr).toInt().coerceAtLeast(0)
    }
}

File: apps/web/src/lib/inpatient-iv-catalog.ts
/** IV fluids and additives commonly available at Ugandan HC III level. */

export type IvFluidOption = { id: string; label: string }
export type IvAdditiveOption = { id: string; label: string; fluidTypes?: string[] }

export const IV_FLUIDS: IvFluidOption[] = [
  { id: 'normal_saline', label: 'Normal saline 0.9%' },
  { id: 'ringers_lactate', label: "Ringer's lactate (Hartmann's)" },
  { id: 'd5', label: 'Dextrose 5%' },
  { id: 'd10', label: 'Dextrose 10%' },
  { id: 'half_normal', label: 'Half-normal saline 0.45%' },
  { id: 'd5_ns', label: 'D5 in normal saline' },
]

export const IV_ADDITIVES: IvAdditiveOption[] = [
  { id: 'none', label: 'None (fluid only)' },
  { id: 'vit_b_complex', label: 'Vitamin B complex' },
  { id: 'vit_c', label: 'Vitamin C' },
  { id: 'ceftriaxone', label: 'Ceftriaxone' },
  { id: 'metronidazole', label: 'Metronidazole' },
  { id: 'artesunate', label: 'Artesunate (severe malaria)' },
  { id: 'quinine', label: 'Quinine' },
  { id: 'oxytocin', label: 'Oxytocin (PPH)' },
  { id: 'magnesium_sulphate', label: 'Magnesium sulphate' },
  { id: 'other', label: 'Other (specify in notes)' },
]

export const IV_VOLUME_PRESETS_ML = [500, 1000, 2000] as const

export function fluidLabel(id: string): string {
  return IV_FLUIDS.find((f) => f.id === id)?.label ?? id
}

export function additiveLabel(id: string | null): string | null {
  if (!id || id === 'none') return null
  return IV_ADDITIVES.find((a) => a.id === id)?.label ?? id
}

/** Estimate ml remaining from start time and rate (no pump — nurse eyeballs drip). */
export function estimateMlRemaining(
  volumeMl: number,
  rateMlHr: number | null,
  startedAt: string,
  now = Date.now(),
): number | null {
  if (!rateMlHr || rateMlHr <= 0) return null
  const elapsedHr = (now - new Date(startedAt).getTime()) / 3_600_000
  return Math.max(0, Math.round(volumeMl - elapsedHr * rateMlHr))
}

export function hoursRemaining(
  volumeMl: number,
  rateMlHr: number | null,
  startedAt: string,
  now = Date.now(),
): number | null {
  const remaining = estimateMlRemaining(volumeMl, rateMlHr, startedAt, now)
  if (remaining == null || !rateMlHr) return null
  return remaining / rateMlHr
}
